### PR TITLE
Display Emojis, Kaomojis, and symbols while in IME mode

### LINKED ISSUE #4148: Tabbed IME fails to insert symbols with a chinese input method
# Environment

```none
Windows build number: 19041.1
Windows Terminal version (if applicable): `911a9dd`
```

# Steps to reproduce

* Use a `Chinese (Simplified, China)` input method (with a `Microsoft Pinyin` keyboard – the default)
* Open the tabbed IME using <kbd>Win</kbd><kbd>.</kbd>
* Insert a few symbols from the symbol tab
* Defocus/Close the IME
* Bring up the Chinese/Pinyin IME again by pressing any keyboard key
* Insert a Chinese character

# Expected behavior

All characters are inserted instantly into the shell prompt.

# Actual behavior

Characters aren't inserted until you defocus/close the IME and bring it up again by pressing a regular keyboard key. [Demo](https://drive.google.com/file/d/1xh-Y7j4l2Ns1oudhNoGu_3ESSFN4NhPj/view)

### PR DESCRIPTION
**Pull request by leonMSFT**
## Summary of the Pull Request
Currently, while in IME mode, selections with the Emoji/Kaomoji/Symbol Picker (which is brought up with <kbd>win+.</kbd>) are not displayed until the user starts a new composition. This is due to the fact that we hide the TextBlock when we receive a CompositionCompleted event, and we only show the TextBlock when we receive a CompositionStarted event. Input from the picker does not count as a composition, so we were never showing the text box, even if the symbols were thrown into the inputBuffer. In addition, we weren't receiving CompositionStarted events when we expected to.

We should be showing the TextBlock when we receive _any_ text, so we should make the TextBlock visible inside of `TextUpdatingHandler`. Furthermore, some really helpful discussion in #3745 around wrapping the `NotifyTextChanged` call with a `NotifyFocusLeave` and a `NotifyFocusEnter` allowed the control to much more consistently determine when a CompositionStarted and a CompositionEnded.

I've also went around and replaced casts with saturating casts, and have removed the line that sets the `textBlock.Width()` so that it would automatically set its width. This resolves the issue where while composing a sentence, the textBlock would be too small to contain all the text, so it would be cut off, but the composition is still valid and still able to continue.

## PR Checklist
* [x] Closes #4148
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [x] Tests added/passed

## Validation Steps Performed
Tested picking emojis, kaomojis, and symbols with numerous different languages.

### CODE REVIEW

File: src/cascadia/TerminalControl/TSFInputControl.cpp [hunk 1]
```diff
@@ -178,12 +178,9 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
 
         // position textblock to cursor position
         _canvas.SetLeft(_textBlock, clientCursorPos.X);
-        _canvas.SetTop(_textBlock, static_cast<double>(clientCursorPos.Y));
+        _canvas.SetTop(_textBlock, ::base::ClampedNumeric<double>(clientCursorPos.Y));
 
-        // width is cursor to end of canvas
-        _textBlock.Width(200); // TODO GitHub #3640: Determine proper Width
```
[the remark this answers is not in the document]
Comment (leonMSFT) [reply src/cascadia/TerminalControl/TSFInputControl.cpp]: Yup, it'll start writing off the end of the window, which actually is the current behavior. Until text wrap-around is implemented, the composition text will always write off the end of the window. Removing this just allows the user to be able to see as much composition text as the window is wide (or until the user resizes the window to make it wider).

File: src/cascadia/WinRTUtils/inc/Utils.h [hunk 2]
```diff
@@ -28,10 +28,10 @@ inline winrt::Windows::UI::Color ColorRefToColor(const COLORREF& colorref)
...
-    const float scaleLocal = gsl::narrow_cast<float>(scale);
-    rect.X *= scaleLocal;
-    rect.Y *= scaleLocal;
-    rect.Width *= scaleLocal;
-    rect.Height *= scaleLocal;
+    const auto scaleLocal = base::ClampedNumeric<float>(scale);
+    rect.X = base::ClampMul(rect.X, scaleLocal);
+    rect.Y = base::ClampMul(rect.Y, scaleLocal);
+    rect.Width = base::ClampMul(rect.Width, scaleLocal);
+    rect.Height = base::ClampMul(rect.Height, scaleLocal);
```
[the remark this answers is not in the document]
Comment (leonMSFT) [reply src/cascadia/WinRTUtils/inc/Utils.h]: as discussed offline, i guess we actually really don't even need clamp for floats, but i'll just leave as is 🤷‍♂️

File: src/cascadia/TerminalControl/TSFInputControl.cpp [hunk 3]
```diff
@@ -334,6 +332,35 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
+    //   clears the input buffer and text block for the next round of input.
+    //   Then hides the text block control until the next time text received.
+    // Arguments:
+    // - <none>
+    // Return Value:
+    // - <none>
+    void TSFInputControl::_SendAndClearText()
+    {
+        // call event handler with data handled by parent
+        _compositionCompletedHandlers(_inputBuffer);
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalControl/TSFInputControl.cpp]: Dumb q: the textblock is relaid-out between these events, right? When composition "completes" without starting, that is?

Comment (leonMSFT) [reply src/cascadia/TerminalControl/TSFInputControl.cpp]: Yup, the textblock actually seems to get relaid-out constantly between just about all events, such as TextUpdate, CompositionStarted, or CompositionCompleted events.

**Pull request merged**